driver/compmisc.ml

```diff
@@ -57,20 +57,42 @@ let initial_env () =
     ~initially_opened_module
     ~open_implicit_modules:(List.rev !Clflags.open_modules)
 
-let read_color_env () =
+type env_clflag =
+  | F : {
+    parse : string -> 'a option;
+    usage : string;
+    flag : 'a option ref;
+    env_var : string;
+  } -> env_clflag
+
+let read_flag_from_env (F { parse; usage; flag; env_var }) =
   try
-    match Clflags.parse_color_setting (Sys.getenv "OCAML_COLOR") with
+    match parse (Sys.getenv env_var) with
     | None ->
         Location.prerr_warning Location.none
-          (Warnings.Bad_env_variable
-             ("OCAML_COLOR",
-              "expected \"auto\", \"always\" or \"never\""));
-    | Some x -> match !Clflags.color with
-      | None -> Clflags.color := Some x
+          (Warnings.Bad_env_variable (env_var, usage))
+    | Some x -> match !flag with
+      | None -> flag := Some x
       | Some _ -> ()
   with
     Not_found -> ()
 
+let read_clflags_from_env () =
+  List.iter read_flag_from_env [
+    F {
+      parse = Clflags.parse_color_setting;
+      usage = "expected \"auto\", \"always\" or \"never\"";
+      flag = Clflags.color;
+      env_var = "OCAML_COLOR";
+    };
+    F {
+      parse = Clflags.parse_error_style;
+      usage = "expected \"contextual\" or \"classic\"";
+      flag = Clflags.error_style;
+      env_var = "OCAML_ERROR_STYLE";
+    };
+  ]
+
```

Comment:
I'm not fond of your implementation.

- It is complex (a record with an existential type)
- It hides the data and exposes only poorly-typed actions, which makes it fairly inflexible

I would think of the following interface.

```
type 'a env_reader
val color_reader : color env_reader
val error_style_reader : error_style env_reader
val set_from_env : 'a option ref -> 'a env_reader -> unit
val read_all_flags_from_env : unit -> unit
```

Thinking about it more, I don't see a reason to have this in Compmisc, I would move it to `Clflags`.

(This API I proposed does not easily allow compiler-libs users to add their own flags, but I don't think we need to worry about this.)

Reply:
> it would create a circular dependency between Clflags and Location (which is used in set_from_env to display a warning).

What about `(unit, bad_env_variable_payload) result`?

Reply:
Thinking about it more:

1. The `result` type solves the dependency problem but it makes the caller-side code cumbersome. Not so nice.
2. What I really like about the idea of moving to `Clflags` was to have the `env_reader` definitions close to their parse functions, etc. It's not nice that the notion of what are the correct error-style settings have to be spilled across two modules. So maybe an acceptable compromise would be to have `set_var_from_env` in Compenv, and the rest in Clflags. That may require either exposing the record type from Clflags, or an internal `try_set_from_env` function with a more complex calling protocol because it is error-handling-parametric (for example the one suggested above).

Reply:
Armael wrote

> Right. I exposed env_reader, and moved the reader definitions to Clflags. I also reused the usage message in compenv.ml as you pointed out, which is nice.

This is now resolved.